yt_dlp/extractor/common.py

```diff
@@ -3812,6 +3844,42 @@ def _yes_playlist(self, playlist_id, video_id, smuggled_data=None, *, playlist_l
         self.to_screen(f'Downloading {playlist_label}{playlist_id} - add --no-playlist to download just the {video_label}{video_id}')
         return True
 
+    @classmethod
+    def extract_from_webpage(cls, ydl, url, webpage):
+        ie = (cls if isinstance(cls._extract_from_webpage, types.MethodType)
+              else ydl.get_info_extractor(cls.ie_key()))
+        yield from ie._extract_from_webpage(url, webpage) or []
+
+    @classmethod
+    def _extract_from_webpage(cls, url, webpage):
+        for embed_url in orderedSet(
+                cls._extract_embed_urls(url, webpage) or [], lazy=True):
+            yield cls.url_result(embed_url, cls)
+
+    @classmethod
+    def _extract_embed_urls(cls, url, webpage):
+        """@returns all the embed urls on the webpage"""
+        if '_EMBED_URL_RE' not in cls.__dict__:
+            assert isinstance(cls._EMBED_REGEX, (list, tuple))
+            for idx, regex in enumerate(cls._EMBED_REGEX):
+                assert regex.count('(?P<url>') == 1, \
+                    f'{cls.__name__}._EMBED_REGEX[{idx}] must have exactly 1 url group\n\t{regex}'
+            cls._EMBED_URL_RE = tuple(map(re.compile, cls._EMBED_REGEX))
+
+        for regex in cls._EMBED_URL_RE:
+            for mobj in regex.finditer(webpage):
+                embed_url = urllib.parse.urljoin(url, unescapeHTML(mobj.group('url')))
+                if cls._VALID_URL is False or cls.suitable(embed_url):
+                    yield embed_url
+
+    class StopExtraction(Exception):
+        pass
+
+    @classmethod
+    def _extract_url(cls, webpage):  # TODO: Remove
```

Comment:
Some extractors call this. It can be removed once all these are rewritten
```
yt_dlp • extractor\breakcom.py:
  44:         youtube_url = YoutubeIE._extract_url(webpage)

yt_dlp • extractor\brightcove.py:
  404:     def _extract_url(ie, webpage):

yt_dlp • extractor\carambatv.py:
  82:         videomore_url = VideomoreIE._extract_url(webpage)

yt_dlp • extractor\cbslocal.py:
  98:         sendtonews_url = SendtoNewsIE._extract_url(webpage)

yt_dlp • extractor\chilloutzone.py:
  71:             youtube_url = YoutubeIE._extract_url(webpage)

yt_dlp • extractor\cracked.py:
  43:         youtube_url = YoutubeIE._extract_url(webpage)

yt_dlp • extractor\cspan.py:
   90:         ustream_url = UstreamIE._extract_url(webpage)
  166:                 senate_isvp_url = SenateISVPIE._extract_url(webpage)

yt_dlp • extractor\ctsnews.py:
  63:             youtube_url = YoutubeIE._extract_url(page)

yt_dlp • extractor\footyroom.py:
  48:             streamable_url = StreamableIE._extract_url(payload)

yt_dlp • extractor\gameinformer.py:
  13:         # normal Brightcove embed code extracted with BrightcoveNewIE._extract_url
  45:         brightcove_url = self.BRIGHTCOVE_URL_TEMPLATE % brightcove_id if brightcove_id else BrightcoveNewIE._extract_url(self, webpage)

yt_dlp • extractor\gdcvault.py:
  175:         embed_url = KalturaIE._extract_url(start_page)

yt_dlp • extractor\heise.py:
  114:         kaltura_url = KalturaIE._extract_url(webpage)

yt_dlp • extractor\meta.py:
  68:         pladform_url = PladformIE._extract_url(webpage)

yt_dlp • extractor\nbc.py:
  250:             NBCSportsVPlayerIE._extract_url(webpage), 'NBCSportsVPlayer')

yt_dlp • extractor\nexx.py:
  527:         return self.url_result(NexxIE._extract_url(webpage), ie=NexxIE.ie_key())

yt_dlp • extractor\normalboots.py:
  39:         jwplatform_url = JWPlatformIE._extract_url(webpage)

yt_dlp • extractor\nowness.py:
  28:                         bc_url = BrightcoveNewIE._extract_url(self, player_code)

yt_dlp • extractor\nzherald.py:
  68:         bc_url = BrightcoveNewIE._extract_url(self, webpage)

yt_dlp • extractor\rcs.py:
  243:                 emb = RCSEmbedsIE._extract_url(page)

yt_dlp • extractor\ukcolumn.py:
  59:         ie, video_url = YoutubeIE, YoutubeIE._extract_url(oembed_webpage)
  61:             ie, video_url = VimeoIE, VimeoIE._extract_url(url, oembed_webpage)

yt_dlp • extractor\vesti.py:
  114:         rutv_url = RUTVIE._extract_url(page)

yt_dlp • extractor\vice.py:
  304:         vice_url = ViceIE._extract_url(body)
  314:         youtube_url = YoutubeIE._extract_url(body)

yt_dlp • extractor\vimeo.py:
  844:             url = self._extract_url(url, self._download_webpage(url, video_id))

yt_dlp • extractor\vk.py:
  390:         youtube_url = YoutubeIE._extract_url(info_page)
  394:         vimeo_url = VimeoIE._extract_url(url, info_page)
  398:         pladform_url = PladformIE._extract_url(info_page)
  413:         odnoklassniki_url = OdnoklassnikiIE._extract_url(info_page)
```

Reply:
but these are extractor-specific apis, not an official api?

Reply:
Yes, which is why I said "It can be removed once all these are rewritten". It is not being kept for backward compat but only to keep everything working till these function calls are updated